_posts/2021-03-19-forecasting-time-series-with-torch_4/forecasting_4.Rmd

```diff
@@ -0,0 +1,570 @@
+
+```{r}
+
+n_timesteps <- 7 * 2
+n_forecast <- 7 * 2
+
+elec_dataset <- dataset(
+  name = "elec_dataset",
+  
+  initialize = function(x, n_timesteps, sample_frac = 1) {
+    
+    self$n_timesteps <- n_timesteps
+    self$x <- torch_tensor((x - train_mean) / train_sd)
+    
+    n <- length(self$x) - self$n_timesteps - 1
+    
+    self$starts <- sort(sample.int(
+      n = n,
+      size = n * sample_frac
+    ))
+    
+  },
+  
+  .getitem = function(i) {
+    
+    start <- self$starts[i]
+    end <- start + self$n_timesteps - 1
+    lag <- 1
+    
+    list(
+      x = self$x[start:end],
+      y = self$x[(start+lag):(end+lag)]$squeeze(2)
+    )
+    
+  },
+  
+  .length = function() {
+    length(self$starts) 
+  }
+)
+
+batch_size <- 32
+
+train_ds <- elec_dataset(elec_train, n_timesteps, sample_frac = 0.5)
+train_dl <- train_ds %>% dataloader(batch_size = batch_size, shuffle = TRUE)
+
+valid_ds <- elec_dataset(elec_valid, n_timesteps, sample_frac = 0.5)
+valid_dl <- valid_ds %>% dataloader(batch_size = batch_size)
+
+test_ds <- elec_dataset(elec_test, n_timesteps)
+test_dl <- test_ds %>% dataloader(batch_size = 1)
+
+```
+
+# Model
+
+Model-wise, we again encounter the three *modules* familiar from the previous post: encoder, decoder, and top-level seq2seq module. Now, however, there is an additional component: the *attention* module, used by the decoder to obtain *attention weights*.
+
+### Encoder
+
+The encoder still works the same way. It basically just wraps an RNN, and returns the final state of the same.
+
+```{r}
+encoder_module <- nn_module(
+  
+  initialize = function(type, input_size, hidden_size, num_layers = 1, dropout = 0) {
+    
+    self$type <- type
+    
+    self$rnn <- if (self$type == "gru") {
+      nn_gru(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        dropout = dropout,
+        batch_first = TRUE
+      )
+    } else {
+      nn_lstm(
+        input_size = input_size,
+        hidden_size = hidden_size,
+        num_layers = num_layers,
+        dropout = dropout,
+        batch_first = TRUE
+      )
+    }
+    
+  },
+  
+  forward = function(x) {
+    
+    x <- self$rnn(x)
+    
+    # return last states for all layers
+    # per layer, a single tensor for GRU, a list of 2 tensors for LSTM
+    x <- x[[2]]
+    x
+    
+  }
+)
+```
+
+### Attention module
+
+In basic seq2seq, whenever it had to generate a new value, the decoder took into account two things: its prior state, and the previous output generated. In an attention-enriched setup, it additionally receives the complete output from the encoder. In deciding what subset of that output should matter right now, it gets help from a new agent, the attention module.
```

Comment:
```suggestion
In basic seq2seq, whenever it had to generate a new value, the decoder took into account two things: its prior state, and the previous output generated. In an attention-enriched setup, the decoder additionally receives the complete output from the encoder. In deciding what subset of that output should matter right now, it gets help from a new agent, the attention module.
```

Reply:
…should matter right now…

What do you mean by "right now" here? As I read the sentence, I don't think you need it, but I could be missing something.

Reply:
no, I think you're right, will remove